pkg/job/discovery.go

```diff
@@ -0,0 +1,281 @@
+				for _, MetricDataResult := range data.MetricDataResults {
+					getMetricData, err := findGetMetricDataByID(input, *MetricDataResult.Id)
+					if err == nil {
+						if len(MetricDataResult.Values) != 0 {
+							getMetricData.GetMetricDataPoint = MetricDataResult.Values[0]
+							getMetricData.GetMetricDataTimestamps = MetricDataResult.Timestamps[0]
+						}
+						output = append(output, &getMetricData)
+					}
+				}
+				mux.Lock()
+				cw = append(cw, output...)
+				mux.Unlock()
+			}
+		}(i)
+	}
+
+	wg.Wait()
+	return resources, cw
+}
+
+func getMetricDataInputLength(metrics []*config.Metric) int64 {
+	var length int64
+	for _, metric := range metrics {
+		if metric.Length > length {
+			length = metric.Length
+		}
+	}
+	return length
+}
+
+func findGetMetricDataByID(getMetricDatas []model.CloudwatchData, value string) (model.CloudwatchData, error) {
+	var g model.CloudwatchData
+	for _, getMetricData := range getMetricDatas {
+		if *getMetricData.MetricID == value {
+			return getMetricData, nil
+		}
+	}
+	return g, fmt.Errorf("metric with id %s not found", value)
+}
+
+func getMetricDataForQueries(
+	ctx context.Context,
+	discoveryJob *config.Job,
+	svc *config.ServiceConfig,
+	region string,
+	accountID *string,
+	tagsOnMetrics model.ExportedTagsOnMetrics,
+	clientCloudwatch *apicloudwatch.CloudwatchInterface,
+	resources []*model.TaggedResource,
+	tagSemaphore chan struct{},
+	logger logging.Logger,
+) []model.CloudwatchData {
+	var getMetricDatas []model.CloudwatchData
+
+	// For every metric of the job
+	for _, metric := range discoveryJob.Metrics {
+		// Get the full list of metrics
+		// This includes, for this metric the possible combinations
+		// of dimensions and value of dimensions with data
+		tagSemaphore <- struct{}{}
+
+		metricsList, err := clientCloudwatch.ListMetrics(ctx, svc.Namespace, metric)
+		<-tagSemaphore
+
+		if err != nil {
+			logger.Error(err, "Failed to get full metric list", "metric_name", metric.Name, "namespace", svc.Namespace)
+			continue
+		}
+
+		if len(resources) == 0 {
+			logger.Debug("No resources for metric", "metric_name", metric.Name, "namespace", svc.Namespace)
+		}
+		getMetricDatas = append(getMetricDatas, getFilteredMetricDatas(logger, region, accountID, discoveryJob.Type, discoveryJob.CustomTags, tagsOnMetrics, svc.DimensionRegexps, resources, metricsList.Metrics, discoveryJob.DimensionNameRequirements, metric)...)
+	}
+	return getMetricDatas
+}
+
+func getFilteredMetricDatas(logger logging.Logger, region string, accountID *string, namespace string, customTags []model.Tag, tagsOnMetrics model.ExportedTagsOnMetrics, dimensionRegexps []*regexp.Regexp, resources []*model.TaggedResource, metricsList []*cloudwatch.Metric, dimensionNameList []string, m *config.Metric) (getMetricsData []model.CloudwatchData) {
+	type filterValues map[string]*model.TaggedResource
+	dimensionsFilter := make(map[string]filterValues)
+	for _, dimensionRegexp := range dimensionRegexps {
```

Comment:
I have a refactor for this logic that is a little bit hard to digest. I will push after this PR, since this is awesome work 😄